core/src/main/java/org/elasticsearch/common/limit/Reservation.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.limit;
+
+/**
+ * An opaque representation of a number of bytes that have been reserved at a limiter. It allows to return the reserved bytes easily (i.e.
+ * without having to remember how many bytes have been reserved on which limiter).
+ */
+public final class Reservation {
+    private final Limiter limiter;
+    private final long bytesReserved;
+
+    public Reservation(Limiter limiter, long bytesReserved) {
+        assert bytesReserved >= 0 : "Reserved bytes: [" + bytesReserved + "] must be >= 0";
+        this.limiter = limiter;
+        this.bytesReserved = bytesReserved;
+    }
+
+
+    public Limiter getLimiter() {
+        return limiter;
+    }
+
+    public long getBytesReserved() {
+        return bytesReserved;
+    }
+
+    /**
+     * Returns the reservation, i.e. it reduces the number of reserved bytes on the correct limiter.
+     */
+    public void doReturn() {
```

Comment:
Hmm.. what about making `Reservation` implement `Closeable` and then releasing them in the `close()` method? This would allow them to be used in try-with-resources